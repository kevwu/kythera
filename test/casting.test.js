const t = require("./util").test

describe("Type casting", () => {
	// TODO type casting is about to be completely overhauled so these tests are disabled
	describe("Custom type assignment", () => {
		t("primitive", `let myIntType = int`, {skip: true})
		t("fn type", `let myFnType = fn<int, str, myIntType> str`, {skip: true})
		t("obj type", `
let myObjType = obj{
    int a,
    str b,
}
	`, {skip: true})
	})

	/*
	t("Casting object", `
let myObj = {
    a = 99,
    b = "beep",
}

myObj.a

(myObj as myObjType).a

myObj.a + 10

let myCastedObj = myObj as myObjType
typeof myCastedObj

typeof myCastedObj == typeof myObj
typeof myCastedObj == typeof (myObj as myObjType)	
	`, {skip: true})
	*/

	/*
	t("Usage as fn param", `
let myObjType = obj{
    int a,
    str b,
}
	
let myObj = {
    a = 99,
    b = "beep",
}

let myCastedObj = myObj as myObjType

// rigid object type use in function
let myFn = <myObjType object> null {
    object.a
    object.b
    return 10
}

myFn(myObj as myObjType)
myFn(myCastedObj)	
	`, {skip: true})
	*/
})
