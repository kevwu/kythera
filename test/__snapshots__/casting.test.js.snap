// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type casting Custom type assignment [PARSE] fn type 1`] = `
Array [
  ParseNode {
    "identifier": "myFnType",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "type",
        "kind": "type",
        "origin": "builtin",
      },
      "value": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
          ParseNode {
            "baseType": "str",
            "kind": "type",
            "origin": "builtin",
          },
          ParseNode {
            "kind": "type",
            "name": "myIntType",
            "origin": "named",
          },
        ],
        "returns": ParseNode {
          "baseType": "str",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
]
`;

exports[`Type casting Custom type assignment [PARSE] obj type 1`] = `
Array [
  ParseNode {
    "identifier": "myObjType",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "type",
        "kind": "type",
        "origin": "builtin",
      },
      "value": ParseNode {
        "baseType": "obj",
        "kind": "type",
        "origin": "builtin",
        "structure": Object {
          "a": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
          "b": ParseNode {
            "baseType": "str",
            "kind": "type",
            "origin": "builtin",
          },
        },
      },
    },
  },
]
`;

exports[`Type casting Custom type assignment [PARSE] primitive 1`] = `
Array [
  ParseNode {
    "identifier": "myIntType",
    "kind": "let",
    "value": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "baseType": "type",
        "kind": "type",
        "origin": "builtin",
      },
      "value": ParseNode {
        "baseType": "int",
        "kind": "type",
        "origin": "builtin",
      },
    },
  },
]
`;
